client/storage-monitor/src/lib.rs

```diff
@@ -0,0 +1,149 @@
+
+	/// How often available space is polled.
+	#[arg(long = "db-storage-polling-period", value_name = "SECONDS", default_value_t = 5, value_parser = clap::value_parser!(u32).range(1..))]
+	pub polling_period: u32,
+}
+
+/// Storage monitor service: checks the available space for the filesystem for fiven path.
+pub struct StorageMonitorService {
+	/// watched path
+	path: PathBuf,
+	/// number of megabytes that shall be free and available on the filesystem for watched path
+	threshold: u64,
+	/// storage space polling period (seconds)
+	pub polling_period: u32,
+}
+
+impl StorageMonitorService {
+	/// Creates new StorageMonitorService for given client config
+	pub fn try_spawn(
+		parameters: StorageMonitorParams,
+		database: DatabaseSource,
+		spawner: &impl SpawnEssentialNamed,
+	) -> Result<(), Error> {
+		Ok(match (parameters.threshold, database.path()) {
+			(0, _) => {
+				log::info!(
+					target: LOG_TARGET,
+					"StorageMonitorService: threshold 0 given, storage monitoring disabled",
+				);
+			},
+			(_, None) => {
+				log::warn!(
+					target: LOG_TARGET,
+					"StorageMonitorService: no database path to observe",
+				);
+			},
+			(threshold, Some(path)) => {
+				log::debug!(
+					target: LOG_TARGET,
+					"Initializing StorageMonitorService for db path: {:?}",
+					path,
+				);
+
+				Self::check_free_space(&path, threshold)?;
+
+				let storage_monitor_service = StorageMonitorService {
+					path: path.to_path_buf(),
+					threshold,
+					polling_period: parameters.polling_period,
+				};
+
+				spawner.spawn_essential(
+					"storage-monitor",
+					None,
+					Box::pin(storage_monitor_service.run()),
+				);
+			},
+		})
+	}
+
+	/// Main monitoring loop, intended to be spawned as essential task. Quits if free space drop
+	/// below threshold.
+	async fn run(self) {
+		loop {
+			tokio::time::sleep(Duration::from_secs(self.polling_period.into())).await;
+			if Self::check_free_space(&self.path, self.threshold).is_err() {
+				break
+			};
+		}
+	}
+
+	/// Returns free space in MB, or error if statvfs failed.
+	fn free_space(path: &Path) -> Result<u64, Errno> {
+		let fs_stats = statvfs(path);
+		fs_stats.map(|stats| stats.blocks_available() * stats.block_size() / 1_000_000)
+	}
+
+	/// Checks if the amount of free space for given `path` is above given `threshold`.
+	/// If it dropped below, error is returned.
+	/// System errors are silently ignored.
+	fn check_free_space(path: &Path, threshold: u64) -> Result<(), Error> {
+		match StorageMonitorService::free_space(path) {
+			Ok(available_space) => {
+				log::trace!(
+					target: LOG_TARGET,
+					"free:{:?} , threshold:{:?}",
+					available_space,
+					threshold,
+				);
+
+				if available_space < threshold {
+					let msg = format!(
+								"Available space {:?}MB for path {:?} dropped below threshold:{:?}MB , terminating...",
+								available_space,
+								path,
+								threshold,
+							);
+					log::error!(target: LOG_TARGET, "{}", msg);
+					Err(msg)
```

Comment:
Not sure it makes that much sense to return the same message here again? We already printed it.

Reply:
Slightly reworked in https://github.com/paritytech/substrate/pull/13082/commits/0b8697e9a9443fa689a14edd99ae8264c2b8ad3b

I introduced `sc_storage_monitor::Error`, and left the log message to be more communicative.